### DISCUSSION TITLE
Register Telerik UI for Blazor

### DISCUSSION THREAD

**Discussion by nicpitsch** [oqtane/oqtane.framework#690] [General] [answered]
1. Has Telerik UI for Blazor to be reigsterd in Oqtane according to Telerik [Client-side UI](https://docs.telerik.com/blazor-ui/getting-started/client-blazor) or [Server-side UI](https://docs.telerik.com/blazor-ui/getting-started/server-blazor) instruction?
1. May Telerik UI for Blazor be registerd in the Oqtane solution or within an external Module?

**Comment by PizzaConsole** [comment]
You can probably follow the steps that @mikecasas did [HERE](https://github.com/oqtane/oqtane.framework/discussions/693)

**Comment by nicpitsch** [comment]
Thanks @PoisnFang, using the IClientStartup interface allows to [register the Telerik Blazor Service, step 6](https://docs.telerik.com/blazor-ui/getting-started/server-blazor#enable-the-components-in-the-project) within an Oqtane Module, Client project like this:

```c#
public class ModuleStartup : IClientStartup
{
	public void ConfigureServices(IServiceCollection services)
	{
		services.AddTelerikBlazor();
	}
}
```

With this it isn't necessary to install Telerik within Oqtane Framework solution.

**Reply by mikecasas** [reply]
@nicpitsch Where did you put the js references? Did you add them to _Host or did you add them in your module?

```
<script src="_content/Telerik.UI.for.Blazor/js/telerik-blazor.js" defer></script>

  <!-- For Trial licenses use
    <script src="_content/Telerik.UI.for.Blazor.Trial/js/telerik-blazor.js" defer></script>
  -->
```

**Comment by nicpitsch** [comment]
@mikecasas first I added the Telerik resources to the index.razor of the module using the List property of *ModuleBase*:

```c#
public override List<Resource> Resources => new List<Resource>()
{
	new Resource { ResourceType = ResourceType.Stylesheet, Bundle="style", Url = ModulePath() + "telerik/all.css"},
	new Resource { ResourceType = ResourceType.Stylesheet, Url = ModulePath() + "Module.css" },
	new Resource { ResourceType = ResourceType.Script, Url = ModulePath() + "telerik/telerik-blazor.js" }
};
```

Lastly it turned out that loading telerik-blazor.js from here seems to be to late, look at [Could not find 'TelerikBlazor' in 'window'](https://docs.telerik.com/blazor-ui/troubleshooting/js-errors). 

Therefore I'm adding the Telerik resources now to *_Host.cshtml* in Oqtane.Server...:

```c#
    ...
    <script src="js/interop.js"></script>

    @* Load Telerik, put resources here with debug.cmd respectively release.cmd  *@
    <script src="Modules/[my-module-name]/telerik/telerik-blazor.js"></script>
    ...
```

...Which means I had to give up the principle not to touch the Oqtane Framework solution.

Maybe there is a better solution, who knows..

**Reply by mikecasas** [reply]
Yes, same issue I am having...I looked back at the issue with quill in the discussions, but I could not get Radzen to work either without putting their js file in the host.cshtml.

**Comment by nicpitsch** [comment]
Next episode.. 
Adding the Telerik resources to _Host.cshtml as I described above is not ideal, because it touches the Oqtane Framework and has some quirks like Telerik styling which gets overwritten. I wanted to follow again the principle **not to touch the Oqtane Framework** solution. It was achieved by applying the theming capabilities of Oqtane. 

This is how I have integrated *Telerik UI for Blazor* now. The steps mentioned below refer basically to the [Telerik instruction, starting at Step 2](https://docs.telerik.com/blazor-ui/getting-started/server-blazor#step-2---enable-the-telerik-components-in-an-existing-project):

## Module

Following the adaptations to apply to the Module projects.

### [ModuleName].Server

Create a folder under *[ModuleName].Server.wwwroot.Themes* with the same name as your Theme namespace and place there the Telerik resources (all.css and telerik-blazor.js). In my example I have named the Theme *TelerikTheme* and placed the resources in a subfolder named *telerik*. The namespace is *Oqtane.Themes.TelerikTheme*.

### [ModuleName].Client

* Install *Telerik.UI.for.Blazor* package (number 1 and 2 of the Telerik instruction).
* Integrate Telerik resources through theming (number 3 and 4):
  * Copy *OqtaneTheme* folder From the Oqtane.Client to the Module Client project and rename it and also the namespaces to *TelerikTheme* (or whatever name you prefer). That's the name for your Theme.
  * Add Telerik stylesheet and JavaScript resources into the *Default.razor* component of the Theme:
```c#
    ...
    // Telerik stylsheet
    new Resource { ResourceType = ResourceType.Stylesheet, Url = ThemePath() + "telerik/all.css" },
    ...
    // Telerik JavaScript    
    new Resource { ResourceType = ResourceType.Script, Bundle = "Bootstrap", Url = ThemePath() + "telerik/telerik-blazor.js" }
```
* Add a *ModuleStartup.cs* class (number 5): 
```c#
    public class ModuleStartup : IClientStartup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddTelerikBlazor();
        }
    }
```
### [ModuleName].Package

* Add xcopy commands to *debug.cmd* (and also *release.cmd*) to get the Telerik Libraries and resources copied to the Oqtane solution:
```
XCOPY "..\Client\bin\Debug\netstandard2.1\Telerik.*.dll" "..\..\[Oqtane]\Oqtane.Server\bin\Debug\netcoreapp3.1\" /Y
XCOPY "..\Server\wwwroot\Modules\[ModuleName]\*" "..\..\[Oqtane]\Oqtane.Server\wwwroot\Modules\pi.ProjAdmin\" /Y /S /I
XCOPY "..\Server\wwwroot\Themes\Oqtane.Themes.TelerikTheme\*" "..\..\[Oqtane]\Oqtane.Server\wwwroot\Themes\Oqtane.Themes.TelerikTheme\" /Y /S /I
```

## Conclusion

[Module_TelerikTheme_example.zip](https://github.com/oqtane/oqtane.framework/files/5117407/Module_TelerikTheme_example.zip) contains the theming based folder structure as described obove. It's taken from the Oqtane Framework theming structure.

Now Build your Module and start Oqtane. Oqtane *Site Settings* lists now the new *Telerik Theme*:
![2020-08-24_11-22-08](https://user-images.githubusercontent.com/1652835/91027366-4cba0480-e5fc-11ea-9ac1-60e1b9ff12b5.png)

It turns out that the Oqtane Framework picks up the Module Theme from i'ts assembly (look at *ThemeRepository.cs*) and treats it like a 'real' Theme - that's so fascinating!

I have no idea if this was that last episode... but so far the Telerik components behave now as expected. 

Maybe somebody has some advices how to do it better.

**Reply by mikecasas** [reply]
So you removed all the references from _Host, and you are not getting the ```Could not find in window error```? I followed your steps, but I am still getting that error.

**Reply by nicpitsch** [reply]
> So you removed all the references from _Host, and you are not getting the `Could not find in window error`? I followed your steps, but I am still getting that error.

@mikecasas yes, I reverted back my changes in *_Host.cshtml*, it's again original Oqtane. I had the *Could not find 'TelerikBlazor' in 'window'* error too. It only went away after adding it to the same Bundle as the oter js: `new Resource..., Bundle = "Bootstrap"...` (look at my example above). I don't understand this behaviour, maybe a bug in Oqtane?

**Reply by sbwalker** [reply]
If Telerik UI for Blazor depends on Bootstrap then you need to declare this dependency by specifying the "bundle" property. This is the way that LoadJS is able to recognize the dependency and load the JS files in the appropriate order.

**Reply by nicpitsch** [reply]
@sbwalker If I don't define a "Bundle" property for that  mentioned *telerik-blazor.js* script or if I set "Bundle" to another name, the script doesn't get loaded at all - means it doesn't not appear in DevTools Network listing. In either case I would have expected the script to be loaded. That's the behavior I don't understand.

**Reply by sbwalker** [reply]
@nicpitsch that does sound like a bug - the Bundle property is supposed to be optional. I will take a closer look at this.

**Reply by sbwalker** [reply]
Note that the example module https://github.com/oqtane/DNF.Projects uses the Bundle property for loading chart.js - and it specifies Bundle = "ChartJS" and it works fine.

**Reply by nicpitsch** [reply]
FYI *telerik-blazor.js* has a size of 1.15 MB, if this could play a role. My Resources in *Default.razor*, the way it works:

```c#
public override List<Resource> Resources => new List<Resource>()
{
	new Resource { ResourceType = ResourceType.Stylesheet, Url = "https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css", Integrity = "sha384-JcKb8q3iqJ61gNV9KGb8thSsNjpSL0n8PARn9HuZOnIxN0hoP+VmmDGMN5t9UJ0Z", CrossOrigin = "anonymous" },
	new Resource { ResourceType = ResourceType.Stylesheet, Url = ThemePath() + "Theme.css" },
	// Telerik stylesheet
	new Resource { ResourceType = ResourceType.Stylesheet, Url = ThemePath() + "telerik/all.css" },
	new Resource { ResourceType = ResourceType.Script, Bundle = "Bootstrap", Url = "https://code.jquery.com/jquery-3.5.1.slim.min.js", Integrity = "sha384-DfXdz2htPH0lsSSs5nCTpuj/zy4C+OGpamoFVy38MVBnE+IbbVYUew+OrCXaRkfj", CrossOrigin = "anonymous" },
	new Resource { ResourceType = ResourceType.Script, Bundle = "Bootstrap", Url = "https://cdn.jsdelivr.net/npm/popper.js@1.16.0/dist/umd/popper.min.js", Integrity = "sha384-Q6E9RHvbIyZFJoft+2mJbHaEWldlvI9IOYy5n3zV9zzTtmI3UksdQRVvoxMfooAo", CrossOrigin = "anonymous" },
	new Resource { ResourceType = ResourceType.Script, Bundle = "Bootstrap", Url = "https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js", Integrity = "sha384-B4gt1jrGC7Jh4AgTPSdUtOBvfO8shuf57BaghqFfPlYxofvL8/KUEfYiJOMMV+rV", CrossOrigin = "anonymous" },
	// Telerik JavaScript    
	new Resource { ResourceType = ResourceType.Script, Bundle = "Bootstrap", Url = ThemePath() + "telerik/telerik-blazor.js" }
};
```

**Reply by sbwalker** [reply]
Those declarations look fine... but are you actually using the Telerik components in your theme? Or are you using them in your custom modules? I am asking because Oqtane has a composite UI where all elements are components that are loaded asynchronously. So if you are declaring Telerik in your theme but trying to use it in a module it is possible that that module is actually loaded BEFORE the theme - which means the Javascript reference would be missing. I realize this is difficult to comprehend at first but there is not a prescriptive hierarchy when loading components like you would imagine based on the logical model.

**Reply by nicpitsch** [reply]
No, I'm not using Telerik components in my theme, only in my Module Client project. The theming is "integrated" into the Module using same structure as Oqtane Framework. This is the folder structure of my *pi.ProjAdmin* Module solution (red=Telerik dependencies, green="integrated" theming):
![piProjAdmin_Module](https://user-images.githubusercontent.com/1652835/91184290-c4675c80-e6ec-11ea-8599-7d02eb6958e6.png)

**Reply by sbwalker** [reply]
The loading behavior of the framework is completely independent from the physical organization of your project. So although your theme and modules are part of the same DLL which is deployed into Oqtane, when a page is composited at run-time it based on individual components - and those components are all loaded asynchronously, there is not a hierarchical loading order. So like I said it is possible that a module component may load before a theme component - this is because they are loaded in parallel and one may load more quickly than another based on their internal logic. Obviously this can cause race conditions to occur because the dependencies have not been loaded. This is especially true for JavaScript. The recommended approach to deal with this issue is to declare your dependencies within the component where they are required. So if your module uses Telerik then you should declare the JavaScript dependencies within the module component - not the theme component. This might seem redundant as instead of using a global declaration you need to repeat this logic within your various components ( or use a base class ) - however it is the best way to encapsulate your components and ensure consistent run-time behavior.

**Reply by nicpitsch** [reply]
> @sbwalker: The loading behavior of the framework is completely independent from the physical organization of your project. So although your theme and modules are part of the same DLL which is deployed into Oqtane, when a page is composited at run-time it based on individual components - and those components are all loaded asynchronously, there is not a hierarchical loading order...

@sbwalker thanks for the explication above and @mikecasas for your reply below regarding "hold back" the rendering of the control. The landing page of my Oqtane did not contain Telerik components/modules, that's why the theming approach seemingly worked. As soon I add a Module containing a Telerik component to the landing page, the *Could not find 'TelerikBlazor' in 'window'* error raises again.

**Comment by mikecasas** [comment]
I'm using Radzen, but the comment above was what helped me. It seemed that the control was always rendering before the script, so I added the below and it worked. Mostly borrowed from the DNF.Projects. I guess the key was to have a property to "hold back" the rendering of the control until the script was loaded.
```
@if (_initialized)
{
    <div class="row">
        <div class="col-md-12">
            <RadzenScheduler @ref="scheduler" style="height: 768px;" TItem="Appointment" Data="@appointments" StartProperty="Start" EndProperty="End"
                             TextProperty="Text" SelectedIndex="2">
                <RadzenDayView />
                <RadzenWeekView />
                <RadzenMonthView />
            </RadzenScheduler>
        </div>
    </div>
}

```

```
protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        try
        {
            await base.OnAfterRenderAsync(firstRender);
            _initialized = true;
        }
        catch (Exception ex)
        {
          
        }
    }


```

**Comment by sbwalker** [comment]
I investigated this further... specifically related to the assertion that there is a bug in the JavaScript registration.

I created a new module using Module Creator. I added a simple function to module.js:

```
var MyCompany = MyCompany || {};

MyCompany.Test = {
    showAlert: function (message) {
        alert(message);
    }
}; 
```

I added an Interop.cs class to the Client project:

```
namespace MyCompany.Test
{
    public class Interop
    {
        private readonly IJSRuntime _jsRuntime;

        public Interop(IJSRuntime jsRuntime)
        {
            _jsRuntime = jsRuntime;
        }

        public Task ShowAlert(string message)
        {
            try
            {
                _jsRuntime.InvokeVoidAsync(
                    "MyCompany.Test.showAlert",
                    message);
                return Task.CompletedTask;
            }
            catch
            {
                return Task.CompletedTask;
            }
        }
    }
}
```

and I included a reference to JavaScript file in the Index.razor ( notice that I did not use the Bundle property as this JavaScript file has no dependencies ):

```
    public override List<Resource> Resources => new List<Resource>()
    {
        new Resource { ResourceType = ResourceType.Script, Url = ModulePath() + "Module.js" }
    };
```

and I wired up a button to call a method which invokes the function:

```
    private async Task ShowAlert()
    {
        var interop = new Interop(JSRuntime);
        await interop.ShowAlert("Test");
    }
```

When I add this module to a page in Oqtane and click the button, the JavaScript alert is successfully displayed. I also used the browser tools to inspect the page DOM and I can see the script reference is properly registered.

```
<script src="Modules/MyCompany.Test/Module.js"></script>
```

**Reply by mikecasas** [reply]
Most of these Blazor libraries (MatBlazor, Radzen, and Telerik) just require the nuget dlls and added references in _Host.cshtml for the JS and/or css. Since we don't want to touch the oqtane framework files, the thought is to just put the references in the module using the standard:

```
 public override List<Resource> Resources => new List<Resource>()
    {
        new Resource { ResourceType = ResourceType.Script, Url = ModulePath() + "3rdPartyJsLib.js" }
        new Resource { ResourceType = ResourceType.StyleSheet, Url = ModulePath() + "3rdPartyCssLib.css" }
    };

```

**So is the pattern to create an Interop class in the module anytime you use a 3rd party Blazor library that needs to reference JavaScript?**

**Reply by mikecasas** [reply]
Sorry just saw PR #711, I guess that is the pattern.

**Reply by sbwalker** [reply]
@mikecasas it is the general pattern for JavaScript... however developers who want to use these larger component libraries will probably want to have them available by default in every module/theme in their installation, so I think an alternate extensibility option to dynamically add them to _host.cshtml during startup is needed.

**Reply by mikecasas** [reply]
I would totally agree.

**Reply by sbwalker** [reply]
I just added support for dynamic inclusion of global resources in _host.cshtml ( ie. global stylesheets and scripts such as those required by UI component suites ) #714

**Reply by sbwalker** [reply]
It works similar to IServerStartup where you can create a class that implements the new IHostResources interface. For consistency it uses the exact same Resource collection/model as what Modules and Themes already use.

**Reply by mikecasas** [reply]
@sbwalker When I try and implement ```IHostResources```, do I also have to inherit from ```ModuleBase``` so I can use ```ModulePath```.

You could hard-code the path, but just wondering the best way.

**Reply by sbwalker** [reply]
The interface is purposely not prescriptive on how to set the path to resources. There are a number of options that come to mind:

- you can manually put the CSS and JS files in the wwwroot/css and wwwroot/js folders and specify the path to these folders in the IHostResources interface
- you can include the CSS and JS files in a custom Theme, include the IHostResources implementation in the Theme assembly, and use ThemePath
- you can include the CSS and JS files in a custom Module, include the IHostResources implementation in the Module assembly, and use ModulePath

**Reply by mikecasas** [reply]
@sbwalker Works perfectly. Thanks again for the improvement!

**Reply by ADefWebserver** [reply]
> I investigated this further... specifically related to the assertion that there is a bug in the JavaScript registration.
> 
> I created a new module using Module Creator. I added a simple function to module.js:
> 
> ```
> var MyCompany = MyCompany || {};
> 
> MyCompany.Test = {
>     showAlert: function (message) {
>         alert(message);
>     }
> }; 
> ```
> 
> I added an Interop.cs class to the Client project:
> 
> ```
> namespace MyCompany.Test
> {
>     public class Interop
>     {
>         private readonly IJSRuntime _jsRuntime;
> 
>         public Interop(IJSRuntime jsRuntime)
>         {
>             _jsRuntime = jsRuntime;
>         }
> 
>         public Task ShowAlert(string message)
>         {
>             try
>             {
>                 _jsRuntime.InvokeVoidAsync(
>                     "MyCompany.Test.showAlert",
>                     message);
>                 return Task.CompletedTask;
>             }
>             catch
>             {
>                 return Task.CompletedTask;
>             }
>         }
>     }
> }
> ```
> 
> and I included a reference to JavaScript file in the Index.razor ( notice that I did not use the Bundle property as this JavaScript file has no dependencies ):
> 
> ```
>     public override List<Resource> Resources => new List<Resource>()
>     {
>         new Resource { ResourceType = ResourceType.Script, Url = ModulePath() + "Module.js" }
>     };
> ```
> 
> and I wired up a button to call a method which invokes the function:
> 
> ```
>     private async Task ShowAlert()
>     {
>         var interop = new Interop(JSRuntime);
>         await interop.ShowAlert("Test");
>     }
> ```
> 
> When I add this module to a page in Oqtane and click the button, the JavaScript alert is successfully displayed. I also used the browser tools to inspect the page DOM and I can see the script reference is properly registered.
> 
> ```
> <script src="Modules/MyCompany.Test/Module.js"></script>
> ```

I posted a Blog post that covers this example here: https://blazorhelpwebsite.com/ViewBlogPost/45

**Comment by sbwalker** [comment]
I have read the documentation for Telerik UI for Blazor and there is a requirement that an application needs to add a TelerikRootComponent to the main Layout file. Oqtane does not use Layouts - it has its own theming engine. I have reached out to Ed Charbonneau to determine oif there an alternate way to register TelerikRootComponent.

**Reply by mikecasas** [reply]
@sbwalker  Will #714 be in a 1.04 release?

**Reply by Mathieu333** [reply]
@sbwalker Hi, I have been playing around with Oqtane and it looks amazing. I am trying to use it with our already existing Telerik projects and I came into the same issue with the missing Layouts and TelerikRootComponent. Did you have an answer from Ed Charbonneau about it? 

Even the Date Picker needs the RootComponent so Telerik cannot be fully used with Oqtane without it. Thank you for your help

**Reply by sbwalker** [reply]
@Mathieu333 I never did receive an answer from Ed Charbonneau... and I have not looked at Telerik since. It seems that other component libraries have simpler integration stories - ie. Syncfusion, Radzen ( https://blazorhelpwebsite.com/ViewBlogPost/47 ). The only suggestion I have if you want to use Telerik is to create a custom Oqtane theme ( you can simply copy the default theme as an example ) and include the TelerikRootComponent as a wrapper around the main body content.

```
@namespace Oqtane.Themes.YourOqtaneTheme
@inherits ThemeBase

<TelerikRootComponent>

<main role="main">
    <nav class="navbar navbar-expand-md navbar-dark bg-primary fixed-top">
        <Logo /><Menu Orientation="Horizontal" />
        <div class="controls ml-md-auto">
            <div class="controls-group"><UserProfile /> <Login /> <ControlPanel /></div>
        </div>
    </nav>
    <div class="content container">
        <PaneLayout />
        <div class="row px-4">
            <Pane Name="Admin" />
        </div>
    </div>
</main>

</TelerikRootComponent>
```

**Reply by Mathieu333** [reply]
Yes I have actually used Radzen today and it seems to work indeed. 
I will try to add the TelerikRootComponent to a custom theme and let everyone know if I succeeded here. 

Thank you, keep up the good work with Oqtane it is exciting 👍

**Reply by Mathieu333** [reply]
Since Telerik added a draggable window and mutliple windows component this might help someone. It is working as intended with a custom Theme and all component can be rendered normaly on Oqtane.

**Comment by sbwalker** [comment]
yes 1.0.4 would be the next release version - are you asking when? I guess the challenge is that if you are building custom solutions then you want to build against official Nuget platform packages rather than the latest code in main?

**Reply by mikecasas** [reply]
Yes, that is true. I can replace the below with file references for the time being.
```
  <ItemGroup>
    <PackageReference Include="Oqtane.Server" Version="1.0.3" />
    <PackageReference Include="Oqtane.Shared" Version="1.0.3" />
  </ItemGroup>
```

**Comment by sbwalker** [comment]
#716 should make it easier to create modules targeting prior versions or the local development environment

**Comment by avazon** [comment]
Does anyone have an updated example for newer Telerik Blazor (version 3.6.1)? I struggle to get it to work but. I think there are breaking changes from version 2.x so I'm doing something wrong when I try to follow the conversation above.
I can get a simple Telerik componont like a Button to work, but the Grid crashes.

**Reply by nicpitsch** [reply]
> Does anyone have an updated example for newer Telerik Blazor (version 3.6.1)? I struggle to get it to work but. I think there are breaking changes from version 2.x so I'm doing something wrong when I try to follow the conversation above. I can get a simple Telerik componont like a Button to work, but the Grid crashes.

@avazon, based on your description the issue is based on data access or data rendering failing.  A simple razor component with Telerik Blazor Grid would look like this:

```cshtml
@namespace pi.MyModule.Modules.piStart
@inherits ModuleBase
@inject IVorstudieService VorstudieService

if (_vorstudies == null)
{
    <p><em>Loading...</em></p>
}
else
{
    @if (_vorstudies.Count != 0)
    {
        <TelerikRootComponent>
            <TelerikGrid Data="@_vorstudies"
                 Pageable="true" PageSize="50" Sortable="true"
                 Resizable="true"
                 Reorderable="true"
                 FilterMode="Telerik.Blazor.GridFilterMode.FilterRow">
                <GridColumns>
                    @*<GridColumn Field="@(nameof(Vorstudie.Id))" Width="140px" />*@
                    <GridColumn Field="@(nameof(Vorstudie.Projektname))" Title="Name" Width="350px" />
                    <GridColumn Field="@(nameof(Vorstudie.Projektnummer))" Title="Nummer" />
                    <GridColumn Title="Vorstudie" Visible="@(UserSecurity.IsAuthorized(PageState.User, PermissionNames.View, ModuleState.Permissions))" Width="130px">
                        <Template Context="vorstudie">
                            <div class="text-center">
                                <MyActionLink Action="Index" ActionType="@ActionTypes.NavigateRight" PageUrlPath="@PageUrlPaths.Vorstudie" Parameters=@($"id={(vorstudie as Vorstudie).Id}") Security="SecurityAccessLevel.View" />
                            </div>
                        </Template>
                    </GridColumn>
                </GridColumns>
            </TelerikGrid>
        </TelerikRootComponent>
    }
}

@code {

    // Variables
    List<Vorstudie> _vorstudies;

    protected override async Task OnInitializedAsync()
    {
        if (PageState.User != null)
        {
            _vorstudies = await VorstudieService.GetUserVorstudiesAsync(ModuleState.ModuleId, PageState.User.UserId);
        }
    }

}
```

**Reply by avazon** [reply]
Thank you, I was unsure where to place the TelerikRootComponent and I think it could have been the problem.

**Comment by nicpitsch** [comment]
For your information, I have changed a bit the Telerik setup (Oqtane version 3.1.4, Blazor Server mode):

**Server Project**

ServerStartup.cs:

```c#
public class ServerStartup : IServerStartup
{
    public IConfigurationRoot Configuration { get; set; }

    public void ConfigureServices(IServiceCollection services)
    {            
        services.AddControllers()
            // Edited: NewtonSoft.Json has been replaced in Oqtane version 3.2.0, look at @sbwalker comment below,
            .AddNewtonsoftJson(options =>
            {
                // NPi, prevent self referencing loop exception when loading related data ('Include' method in EF Core),
                // https://community.inkdrop.app/note/a34367a720ac0ba5ac22bcf2fb149aaf/note:-Y7BmxnAX.
                // Be aware of possible issues by allowing this option (https://dotnetcoretutorials.com/2020/03/15/fixing-json-self-referencing-loop-exceptions/).
                options.SerializerSettings.ReferenceLoopHandling = Newtonsoft.Json.ReferenceLoopHandling.Ignore;
                // NPi, suppress NewtonsoftJason default behaviour setting member names casing to camelCase (https://stackoverflow.com/a/60753119/2349178).
                options.UseMemberCasing();
            });

        // Telerik.UI.for.Blazor is setup in ClientStartup.cs in Client Project.
        //
    }  
}
```

HostResources.cs:

```c#
public class HostResources : IHostResources
{
    // Dynamic inclusion of global resources in _host.cshtml.
    // Oqtane: https://github.com/oqtane/oqtane.framework/discussions/690#discussioncomment-58579
    List<Resource> IHostResources.Resources => new List<Resource>()
    {
        new Resource { ResourceType = ResourceType.Stylesheet, Url = "Modules/pi.MyModule/Telerik.UI.for.Blazor/css/kendo-theme-default/all.css", Location = ResourceLocation.Head, Level = ResourceLevel.App },
        // Alternatively (https://docs.telerik.com/reporting/using-reports-in-applications/display-reports-in-applications/web-application/blazor-report-viewer/how-to-use-blazor-report-viewer#adding-the-html5-report-viewer-component):
        //new Resource { ResourceType = ResourceType.Stylesheet, Url = "https://kendo.cdn.telerik.com/2022.1.119/styles/kendo.common.min.css", Location = ResourceLocation.Head, Level = ResourceLevel.App },
        //new Resource { ResourceType = ResourceType.Stylesheet, Url = "https://kendo.cdn.telerik.com/2022.1.119/styles/kendo.default.min.css", Location = ResourceLocation.Head, Level = ResourceLevel.App },
        //
        new Resource { ResourceType = ResourceType.Stylesheet, Url = "Modules/pi.MyModule/Module.css", Location = ResourceLocation.Head, Level = ResourceLevel.App },
        new Resource { ResourceType = ResourceType.Script, Url = "Modules/pi.MyModule/Telerik.UI.for.Blazor/js/telerik-blazor.js" },            
        new Resource { ResourceType = ResourceType.Script, Url = "Modules/pi.MyModule/Module.js" },
    };
}
```

**Client Project**

ClientStartup.cs:

```c#
public class ClientStartup : IClientStartup
{
    public void ConfigureServices(IServiceCollection services)
    {
        // Telerik UI for Blazor setup
        // (https://docs.telerik.com/blazor-ui/getting-started/what-you-need#project-configuration).
        services.AddTelerikBlazor();
        // END Telerik UI
    }
}
```

**Reply by sbwalker** [reply]
Note that NewtonSoft.Json was removed as a dependency from the core framework in version 3.2.1 (the last release). This was because there was a security vulnerability in the NewtonSoft library - and the goal in Oqtane is to minimize dependencies. The framework now relies solely on System.Text.Json (the native .NET Json serializer). So if Telerik (or any other library) has a dependency on NewtonSoft you will need to package it with your module or theme package.

**Comment by sbwalker** [comment]
FYI... Telerik was kind enough to provide me with a commercial license (a Microsoft MVP benefit) so that I can experiment with the Oqtane integration myself as well.

**Reply by avazon** [reply]
Great! Looking forward to see amazing samples with Telerik. I got it to work with the tip/sample from nicpitsch.
But I think it would be nice to have the TelerikRootComponent in a Theme as discussed so it does not have to be repeated in every module. (I have problems with that).
Also nice if it was possible to use in the Oqtane framework like in the Admin pages (as an option).

**Reply by gpticar-oa** [reply]
@sbwalker Have you got any new info on this implementing Telerik i Oqtane through module?

**Reply by sbwalker** [reply, marked as answer]
@gpticar see https://github.com/nicpitsch/Oqtane.TelerikModule1

**Reply by gpticar-oa** [reply]
@sbwalker Works like a charm, thank you.
